### PR TITLE
BFD-2886: Build Docker Host AMI

### PR DESCRIPTION
**Pull request by billie-rose**
**JIRA Ticket:**
[BFD-2886](https://jira.cms.gov/browse/BFD-2886)

**User Story or Bug Summary:**

As a BFD ops Engineer, I'd like an AMI which runs Docker as a process so that containers can be run on an instance of this AMI.

---

### What Does This PR Do?

Create an AMI with Docker baked in as a service
Note that due to creating this AMI outside of ansible, it is currently **not** possible to SSH into the AMI instance using our standard process. This hinderance is to be solved by [BFD-2952: Refactor SSH Management](https://jira.cms.gov/browse/BFD-2952). 

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] The data dictionary has been updated with any field mapping changes, if any were made.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: ops/jenkins/global-pipeline-libraries/vars/awsEc2.groovy
Comment (mjburling) [inline ops/jenkins/global-pipeline-libraries/vars/awsEc2.groovy:1]: Fantastic!

File: ops/packer/scripts/build_bfd-docker-host.sh [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
```
Comment (mjburling) [inline ops/packer/scripts/build_bfd-docker-host.sh:1]: I believe this is still the preferred, _more_ portable shebang. However, I've recently read that this isn't so cut and dry, and some really **do** prefer the simpler variant that you've offered here for fear of `env` not being available. I'll leave it up to you as to whether you want to follow this or an alternative ¯\\\_(ツ)_/¯ 

```suggestion
#!/usr/bin/env bash
```

File: ops/deploy-ccs.groovy [hunk 2]
```diff
@@ -118,32 +90,61 @@ def buildAppAmis(String gitBranchName, String gitCommitId, AmiIds amiIds, AppBui
...
+ *
+ * @param gitBranchName the name of the Git branch this build is for
+ * @param gitCommitId the hash/ID of the Git commit that this build is for
+ * @param the ID of the base "platinum" AMI to use for new instances/AMIs
+ * @return the AMI id of the Docker Host AMI
+ * @throws RuntimeException An exception will be bubbled up if the AMI-builder tooling returns a non-zero exit code.
+ */
+def buildDockerHostAmi(String gitBranchName, String gitCommitId, String platinumAmiId) {
+	dir('ops/packer'){
+		packerBuildAmis(platinumAmiId, gitBranchName, gitCommitId, "./build_bfd-docker-host.json")
```
Comment (mjburling) [inline ops/deploy-ccs.groovy:121]: We didn't discuss this at length, but I really appreciate you decoupling this from the monolithic packer manifest and encoding it into a distinct, docker-host manifest. 👍

File: ops/deploy-ccs.groovy [hunk 3]
```diff
@@ -107,6 +78,7 @@ def findAmis(String branchName = 'master') {
  * @throws RuntimeException An exception will be bubbled up if the AMI-builder tooling returns a non-zero exit code.
  */
 def buildAppAmis(String gitBranchName, String gitCommitId, AmiIds amiIds, AppBuildResults appBuildResults) {
+	amiIdsWrapper = new AmiIds();
```
Comment (mjburling) [inline ops/deploy-ccs.groovy:81]: 🐶 Much readable.

**Review by mjburling — APPROVED** [review]

File: ops/packer/scripts/build_bfd-docker-host.sh (hunk 1 above)
Comment (billie-rose) [reply ops/packer/scripts/build_bfd-docker-host.sh:1]: I'm tempted to leave this as-is and change it in 2887 so I don't need to re-test the AMI creation (which is probably unnecessary but still...better safe than sorry)

**Review by brandoncruz3 — APPROVED** [review]

**Pull request merged**